### PR TITLE
Mvc delievery mechanism domain model

### PR DESCRIPTION
**Pull request by sandromancuso**

### CODE REVIEW

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+With the Model now associated to Entities and persistency and the View associated to HTML, CSS and JavaScript, developers found no other alternative then to put the business logic of the system in Controller layer. Unfortunately MVC frameworks are one of the main reasons we find Web applications with an anaemic domains (Entities with no behaviour) and fat Controllers - responsible for navigation, managing sessions, parsing JSON and XML, and also with business logic.
```
Comment (alvarogarcia7) [inline _posts/2017-09-20-mvc-delievery-mechanism-dm.md:30]: Maybe throw in a link to Anaemic domain?

You can choose:

- https://en.wikipedia.org/wiki/Anemic_domain_model
- https://www.martinfowler.com/bliki/AnemicDomainModel.html
- https://dzone.com/articles/anaemic-vs-rich-domain-objects-finding-the-balance

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+With the Model now associated to Entities and persistency and the View associated to HTML, CSS and JavaScript, developers found no other alternative then to put the business logic of the system in Controller layer. Unfortunately MVC frameworks are one of the main reasons we find Web applications with an anaemic domains (Entities with no behaviour) and fat Controllers - responsible for navigation, managing sessions, parsing JSON and XML, and also with business logic. 
+
+From the original MVC idea, the Controller layer should be a very thin layer, handling the requests from the View, delegating all the business behaviour to the Model. Depending on the result returned by the Model, the Controller decides which View to display next.  
+
```
Comment (alvarogarcia7) [inline _posts/2017-09-20-mvc-delievery-mechanism-dm.md:33]: I would appreciate a link/note explaining about MVC libraries (e.g. some tools in clojure for web) and why/why not they popular

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+</center>
+<br/> 
+
+Although reducing boiler plate code using frameworks sounds a great idea, we are coupling the structure of our entities (domain model) to the APIs we provide to the external world and our database. Changing the database impacts the API and vice-versa. What had initially been seen as a time saver, now is a reason to avoid change. The larger and more coupled the system grows, the less developers will be inclined to make changes due to the size of the ripple effects. Many web apps today have APIs which are far from ideal as they reflect the internal data structure of the application instead of focusing only on the information that would make more sense to the external world. 
+
+Do not be afraid to create your own mappers. They are normally very simple to write and there are many small libraries specialised to parse objects to JSON or persist data to databases. The advantage of writing our own mappers is that we do not need to couple our APIs or databases to anything. Changes are localised and easy to change. On top of that, we can easily test-drive our mappers and move API tests to the unit level instead of doing it at Acceptance level.
```
Comment (alvarogarcia7) [inline _posts/2017-09-20-mvc-delievery-mechanism-dm.md:111]: +1 to this. Well written

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+**Deployable Domain Model**
+
+<center>
+<img src="{{site.baseurl}}/assets/img/custom/blog/2017-09-20-mvc-delivery-mechanism/deployable-dm.png" alt="Deployable Domain Model" class="img img-responsive" style="height: 70%; width: 70%;"/>
+</center>
+<br/> 
+
+# Summary
+
+In MVC, Controllers should be a very thin layer (2 or 3 lines) which identify the request from the user, invoke the appropriate behaviour in the Model, and invoke the appropriate View once the Model returns. Model is not only about entities (state); it’s about all the behaviour of your system, the entire Domain Model.
```
Comment (alvarogarcia7) [inline _posts/2017-09-20-mvc-delievery-mechanism-dm.md:137]: Disagree with the method size (in line numbers). 

Even if all your business logic is handled in services/model, you need to deal with the conversion from internal (business) errors/messages to HTTP errors + statuses. 

Personally, I'd keep it as 

> very thin layer (to the absolute minimum) which identify

...

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+The choice of View technology impacts on how your layers are coupled and how your application is deployed. 
+
+Keep the Delivery Mechanism decoupled from your Domain Model. The Domain Model should not know anything about how the business  feature and data are delivered to the outside world.
```
Comment (alvarogarcia7) [inline _posts/2017-09-20-mvc-delievery-mechanism-dm.md:141]: > You can even have the layers live in different modules / dlls to enforce this separation

**Review by alvarogarcia7 — COMMENTED** [review]
Good post. Long and interesting

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md (hunk 1 above)
Comment (sandromancuso) [reply _posts/2017-09-20-mvc-delievery-mechanism-dm.md:30]: Good idea. I'll add it

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md (hunk 2 above)
Comment (sandromancuso) [reply _posts/2017-09-20-mvc-delievery-mechanism-dm.md:33]: I'll see if I can fit them in but the post is already long and goes off in a few tangents.

File: _posts/2017-09-20-mvc-delievery-mechanism-dm.md (hunk 4 above)
Comment (sandromancuso) [reply _posts/2017-09-20-mvc-delievery-mechanism-dm.md:137]: Makes sense. the number of lines is not the main point here and will take the focus away from the actual message. I'll change it.

**Pull request merged**